core/src/main/scala/org/bitcoins/core/wallet/signer/Signer.scala

```diff
@@ -8,13 +8,24 @@ import org.bitcoins.core.protocol.transaction._
 import org.bitcoins.core.script.crypto.HashType
 import org.bitcoins.core.script.flag.ScriptFlag
 import org.bitcoins.core.wallet.builder.TxBuilderError
-import org.bitcoins.core.wallet.utxo.UTXOSpendingInfo
+import org.bitcoins.core.wallet.utxo.{
+  BitcoinUTXOSpendingInfo,
+  LockTimeSpendingInfo,
+  MultiSignatureSpendingInfo,
+  P2PKHSpendingInfo,
+  P2PKSpendingInfo,
+  P2SHSpendingInfo,
+  P2WPKHV0SpendingInfo,
+  P2WSHV0SpendingInfo,
+  UTXOSpendingInfo,
+  UnassignedSegwitNativeUTXOSpendingInfo
+}
 import scodec.bits.ByteVector
 
 import scala.concurrent.{ExecutionContext, Future}
 
 /** The class used to represent a signing process for a specific [[org.bitcoins.core.protocol.script.ScriptPubKey]] type */
-sealed abstract class Signer {
+sealed abstract class Signer[-SpendingInfo <: UTXOSpendingInfo] {
```

Comment:
Why are these contravariant? If I remove the `-` from `Signer` and `BitcoinSigner` things still seem to compile

https://docs.scala-lang.org/tour/variances.html

Reply:
If something can sign given a more generic type then it is more specific. Marking it as contra-variant lets us use this fact, and the fact that it compiles with the `-` is just proof that `Signer` is contra-variant in its parameter.

To make things explicit with an example:
`Signer[P2SHNestedSegwitSpendingInfo]` is the type for all things that can sign p2sh(segwit).
`Signer[P2SHNoNestSpendingInfo]` is the type for all things that can sign p2sh(raw).
`Signer[P2SHSpendingInfo]` is the type for all things that can sign both and hence it is a subtype of those two.